pkg/cosign/verify.go

```diff
@@ -611,117 +641,145 @@ func verifySignatures(ctx context.Context, sigs oci.Signatures, h v1.Hash, co *C
 }
 
 // verifyInternal holds the main verification flow for signatures and attestations.
-//  1. Verifies the signature using the provided verifier.
-//  2. Checks for transparency log entry presence:
+//   - Checks for transparency log entry presence:
 //     a. Verifies the Rekor entry in the bundle, if provided. This works offline OR
 //     b. If we don't have a Rekor entry retrieved via cert, do an online lookup (assuming
 //     we are in experimental mode).
-//  3. If a certificate is provided, check it's expiration using the transparency log timestamp.
-func verifyInternal(ctx context.Context, sig oci.Signature, h v1.Hash,
+//   - If a certificate is provided and necessary, validates it against CheckOpts
+//   - Verifies the signature (using the provided verifier, if any).
+func verifyInternal(ctx context.Context, untrustedSignature oci.Signature, h v1.Hash,
 	verifyFn signatureVerificationFn, co *CheckOpts) (
 	bundleVerified bool, err error) {
+	var acceptableRFC3161Time, acceptableRekorBundleTime *time.Time // Timestamps for the signature we accept, or nil if not applicable.
+
+	// === If the policy accepts a RFC 3161 timestamp, see if we have an acceptable one.
+	// A missing timestamp is not an error in itself.
+	if co.TSACerts != nil {
+		acceptableRFC3161Timestamp, err := VerifyRFC3161Timestamp(ctx, untrustedSignature, co.TSACerts)
+		if err != nil {
+			return false, fmt.Errorf("unable to verify RFC3161 timestamp bundle: %w", err)
+		}
+		if acceptableRFC3161Timestamp != nil {
+			bundleVerified = true
+			acceptableRFC3161Time = &acceptableRFC3161Timestamp.Time
+		}
+	}
+
+	// === If the policy requires Rekor presence, enforce that.
+	// Also extract a timestamp of Rekor inclusion.
+	if !co.SkipTlogVerify {
+		bundleVerified, err = VerifyBundle(ctx, untrustedSignature, co)
+		if err != nil {
+			return false, fmt.Errorf("error verifying bundle: %w", err)
+		}
+
+		if bundleVerified {
+			// Update with the verified bundle's integrated time.
+			t, err := getBundleIntegratedTime(untrustedSignature)
+			if err != nil {
+				return false, fmt.Errorf("error getting bundle integrated time: %w", err)
+			}
+			acceptableRekorBundleTime = &t
+		} else {
+			// If the --offline flag was specified, fail here. bundleVerified returns false with
+			// no error when there was no bundle provided.
+			// TODO: You can be offline when you are verifying with a public key. This shouldn't
+			// error out, but maybe should be a log message.
+			if co.Offline {
+				return false, fmt.Errorf("offline verification failed")
+			}
+
+			if co.RekorClient != nil {
```

Comment:
I think this check was always very confusing to me. It seems like if you aren't skipping TLOG verification, then if there's no rekor client, somehow you can still pass without checking the tlog.

do you think maybe it should be an error that co.RekorClient is nil when we have to do tlog verification and never got a bundle?

Reply:
Right now, the API is such that external callers can call `VerifyImageSignature` and get `(bundleVerified = false, err = nil)`. So this PR doesn’t change that, it’s a large enough set of changes already.

I agree that it seems dangerous and undesirable; IMHO external users of `pkg/cosign` should just provide a policy, and not be relied upon to correctly enforce other conditions on top. Conceptually, I think of “change the API philosophy” and “change the implementation philosophy” are fairly separate conversations — OTOH, admittedly, the question of the right timestamp to use for certificate expiry means the two can’t separated that cleanly.